### PR TITLE
Various CI pipeline improvements

### PR DESCRIPTION
**Pull request by pulsastrix**
This PR makes some larger improvements to the `libcoap-rs` CI pipeline:
- Switched from `cargo tarpaulin` to rustc's builtin `-C instrument-coverage` flag+`grcov` for coverage.
- `grcov` generates a report in both Markdown and HTML format, with the Markdown being used for the PR sticky comment and the HTML version (that allows looking at the source code directly) being deployed to GitHub Pages and linked in the sticky comment.
- added a job for `rustfmt`
- switched from [giraffate/clippy-action](https://github.com/giraffate/clippy-action) to [qnighy/clippy-reviewdog-filter](https://github.com/qnighy/clippy-reviewdog-filter) as the former one seems to not always work properly.
- Tests are now run for the Minimum Supported Rust Version, the current stable version and the current nightly release.

### CODE REVIEW

File: libcoap/src/context.rs
Comment (pulsastrix) [reply libcoap/src/context.rs:649]: Left here to illustrate how clippy suggestions look like.

File: libcoap/src/crypto/mod.rs
Comment (pulsastrix) [reply libcoap/src/crypto/mod.rs:73]: Put this in here intentionally to illustrate how rustfmt comments look like.

**Comment by pulsastrix** [comment]
Quick question to the other maintainers @JKRhb @falko17: Do you think it's worth the effort to also build the documentation and push it to GitHub Pages? 

If so, would it make sense to do this for every pull request or only for the `main` branch?

Doing so for every PR might make it easier for people to review documentation changes, but would also cause many different versions of the documentation to be available at the same time.

As of now, I have not implemented a cleanup mechanism to remove old coverage reports (and possible documentation builds) of merged PRs from the GitHub pages deployment. This could be seen as an advantage (you can see the coverage reports even after the PR was merged), but I'm not sure if this could also have negative implications regarding repository size.

**Comment by JKRhb** [comment]
> Quick question to the other maintainers @JKRhb @falko17: Do you think it's worth the effort to also build the documentation and push it to GitHub Pages?
> 
> If so, would it make sense to do this for every pull request or only for the `main` branch?
> 
> Doing so for every PR might make it easier for people to review documentation changes, but would also cause many different versions of the documentation to be available at the same time.

Hmm, I guess it cannot hurt, but I am not sure whether it has that much of a benefit 🤔 If you want to review the rendered version, I guess you could also just quickly build the documentation locally. Then again, it might be nice to have links pointing to places where the rendering has issues 🤔 I guess it comes down to how easy this is to implement.

> As of now, I have not implemented a cleanup mechanism to remove old coverage reports (and possible documentation builds) of merged PRs from the GitHub pages deployment. This could be seen as an advantage (you can see the coverage reports even after the PR was merged), but I'm not sure if this could also have negative implications regarding repository size.

Would it be an option to not deploy GitHub pages via a branch but via an action? I guess then we would not have to worry about repository size at all. Maybe older reports could be stored in the cache for this approach.

**Comment by pulsastrix** [comment]
> Hmm, I guess it cannot hurt, but I am not sure whether it has that much of a benefit 🤔 If you want to review the rendered version, I guess you could also just quickly build the documentation locally. Then again, it might be nice to have links pointing to places where the rendering has issues 🤔 I guess it comes down to how easy this is to implement.

Probably not that complicated, I'll see what I can do.

> Would it be an option to not deploy GitHub pages via a branch but via an action? I guess then we would not have to worry about repository size at all. Maybe older reports could be stored in the cache for this approach.

Deploying via an action would have the disadvantage that we'd only have a single deployment for all branches, i.e., pushing to one of the current branches will override the deployment of the other one.

We could [deploy to a different repository](https://github.com/peaceiris/actions-gh-pages?tab=readme-ov-file#%EF%B8%8F-deploy-to-external-repository-external_repository), but in order to do so, we would have to create an access token, as the standard `GITHUB_TOKEN` cannot access other repositories.

**Comment by pulsastrix** [comment]
On another note, I just noticed that I might have to make some additional changes to properly support PRs from forked repositories, so I'll set this back to the draft state.

**Comment by JKRhb** [comment]
> Deploying via an action would (if I understand it correctly) have the disadvantage that we'd only have a single deployment for all branches, i.e., pushing to one of the current branches will override the deployment of the other one.

Hmm, yeah, I guess in that case we would need to have an action that iterates over all (active) branches/PRs and then create the documentation/coverage reports for all of them. This would probably have the main downside of being slow if there are many branches/PRs.

**Comment by pulsastrix** [comment]
This should do it, sorry for filling your inbox with notifications...

- There is now also a build for the documentation.
- The docs and coverage report are now deployed to GitHub Pages using [a different repository](https://github.com/namib-project/libcoap-rs-docs) to avoid bloating the main repository size.
- I have made sure that builds from forks also work, see namib-project/libcoap-rs#39.
- Linter suggestions are now always provided as GitHub checks and not as review comments (review comments would require access to this repository's secrets earlier on in the pipeline).
- The README files have been updated to include relevant badges.

To test this PR, I had to temporarily change the default branch, as `workflow_run`-triggered runs use the version of the workflow file that is provided on the default branch and not the one of the PR's branch. 
This is a security feature, as the `workflow_run` triggered run of `report.yml` has access to the repository's secrets, see https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/.

File: .github/workflows/report.yml
Comment (falko17) [inline .github/workflows/report.yml:57]: Accidental German:
```suggestion
      # --all-features uses GNUTLS as backend, must provide it.
```

File: .github/workflows/report.yml
Comment (falko17) [inline .github/workflows/report.yml:74]: Just out of interest, does this mean we run the CI pipeline for (external) PRs by default, without any prior approval?

File: libcoap/src/context.rs
Comment (falko17) [inline libcoap/src/context.rs:649]: Just as a reminder, to make sure, we should remember to remove these before merging.

**Review by falko17 — CHANGES_REQUESTED** [review]
Looks good to me! I have just three small comments.

File: .github/workflows/report.yml
Comment (pulsastrix) [reply .github/workflows/report.yml:74]: This is just out of additional caution, external PRs are also set to always require manual approval.

**Review by falko17 — APPROVED** [review]

**Comment by pulsastrix** [comment]
Thanks for the quick review!

Remaining CI pipeline failures are unrelated to this PR (see #38), so I'll go ahead and merge.

**Pull request merged**
